src/main/java/net/neoforged/neoforge/items/VanillaInventoryCodeHooks.java

```diff
@@ -192,19 +187,33 @@ private static boolean isEmpty(IItemHandler itemHandler) {
         return true;
     }
 
-    public static Optional<Pair<IItemHandler, Object>> getItemHandler(Level worldIn, double x, double y, double z, final Direction side) {
-        int i = Mth.floor(x);
-        int j = Mth.floor(y);
-        int k = Mth.floor(z);
-        BlockPos blockpos = new BlockPos(i, j, k);
-        net.minecraft.world.level.block.state.BlockState state = worldIn.getBlockState(blockpos);
+    private static Optional<Pair<IItemHandler, Object>> getAttachedItemHandler(Level level, BlockPos pos, Direction direction) {
+        return getItemHandlerAt(level, pos.getX() + direction.getStepX() + 0.5, pos.getY() + direction.getStepY() + 0.5, pos.getZ() + direction.getStepZ() + 0.5, direction.getOpposite());
+    }
 
-        if (state.hasBlockEntity()) {
-            BlockEntity blockEntity = worldIn.getBlockEntity(blockpos);
-            if (blockEntity != null) {
-                return blockEntity.getCapability(Capabilities.ITEM_HANDLER, side)
-                        .map(capability -> ImmutablePair.<IItemHandler, Object>of(capability, blockEntity));
-            }
+    private static Optional<Pair<IItemHandler, Object>> getSourceItemHandler(Level level, Hopper hopper) {
+        return getItemHandlerAt(level, hopper.getLevelX(), hopper.getLevelY() + 1.0, hopper.getLevelZ(), Direction.DOWN);
+    }
+
+    private static final Predicate<Entity> ITEM_HANDLER_ENTITY_SELECTOR = (entity) -> {
+        return entity.getCapability(Capabilities.ItemHandler.ENTITY_AUTOMATION) != null;
```

Comment:
Do we want to validate the entity is not in spectator? A lot of vanilla's entity predicates filter out spectators, though looking at it vanilla's predicate for it is `instanceof Container && isAlive` so we probably want to validate the entity is alive, even though I think it would make sense from a capability perspective to also filter out spectators (especially as some mods may expose at least some of the player's inventory to that capability).

Reply:
I think it would be the mod's responsibility to make sure that the inventory is not exposed to the automation capability if automation should not touch it 😄

Added the `isAlive` check